src/connection/socket-connection.ts

```diff
@@ -0,0 +1,208 @@
+
+  constructor(
+    private socket: WebSocket,
+    private dwn: Dwn
+  ){
+    socket.on('message', this.message.bind(this));
+    socket.on('close', this.close.bind(this));
+    socket.on('error', this.error.bind(this));
+    socket.on('pong', this.pong.bind(this));
+
+    // Sometimes connections between client <-> server can get borked in such a way that
+    // leaves both unaware of the borkage. ping messages can be used as a means to verify
+    // that the remote endpoint is still responsive. Server will ping each socket every 30s
+    // if a pong hasn't received from a socket by the next ping, the server will terminate
+    // the socket connection
+    socket[SOCKET_ISALIVE_SYMBOL] = true;
+    this.heartbeatInterval = setInterval(() => {
+      if (this.socket[SOCKET_ISALIVE_SYMBOL] === false) {
+        this.close();
+      }
+      this.socket[SOCKET_ISALIVE_SYMBOL] = false;
+      this.socket.ping();
+    }, HEARTBEAT_INTERVAL);
+  }
+
+  /**
+   * Adds a reference for the JSON RPC Subscription to this connection.
+   * Used for cleanup if the connection is closed.
+   */
+  async subscribe(subscription: JsonRPCSubscription): Promise<void> {
+    if (this.subscriptions.has(subscription.id)) {
+      throw new DwnServerError(
+        DwnServerErrorCode.ConnectionSubscriptionJsonRPCIdExists,
+        `the subscription with id ${subscription.id} already exists`
+      )
+    }
+
+    this.subscriptions.set(subscription.id, subscription);
+  }
+
+  /**
+   * Closes and removes the reference for a given subscription from this connection.
+   *
+   * @param id the `JsonRpcId` of the JSON RPC subscription request.
+   */
+  async closeSubscription(id: JsonRpcId): Promise<void> {
+    if (!this.subscriptions.has(id)) {
+      throw new DwnServerError(
+        DwnServerErrorCode.ConnectionSubscriptionJsonRPCIdNotFound,
+        `the subscription with id ${id} was not found`
+      )
+    }
+
+    const connection = this.subscriptions.get(id);
+    await connection.close();
+    this.subscriptions.delete(id);
+  }
+
+  /**
+   * Closes the existing connection and cleans up any listeners or subscriptions.
+   */
+  async close(): Promise<void> {
+    clearInterval(this.heartbeatInterval);
+    // clean up all socket event listeners
+    this.socket.removeAllListeners();
+
+    const closePromises = [];
+    for (const [id, subscription] of this.subscriptions) {
+      closePromises.push(subscription.close());
+      this.subscriptions.delete(id);
+    }
+
+    // close all of the associated subscriptions
+    await Promise.all(closePromises);
+
+    // close the socket.
+    this.socket.close();
+  }
+
+  /**
+   * Pong messages are automatically sent in response to ping messages as required by
+   * the websocket spec. So, no need to send explicit pongs.
+   */
+  private pong(): void {
+    this.socket[SOCKET_ISALIVE_SYMBOL] = true;
+  }
+
+  private async error(error?:Error): Promise<void>{
+    if (error) {
```

Comment:
Sanity: if error event triggered, we'd want to close even if no error param is given no? (though I'd think error would always be given, but that's probably a separate point).

Reply:
I was borrowing this for the send callback as well as the socket error call back. I modified the implementation a bit to allow for an optional send callback that defaults to logging an error if one is returned.